lld/ELF/Writer.cpp

```diff
@@ -920,7 +929,9 @@ static unsigned getSectionRank(const OutputSection &osec) {
     // TLS sections directly before the other RELRO sections.
     if (!(osec.flags & SHF_TLS))
       rank |= RF_NOT_TLS;
-    if (!isRelroSection(&osec))
+    if (isRelroSection(&osec))
+      osec.relro = true;
```

Comment:
IIUC this was previously used for sections between DATA_SEGMENT_ALIGN and DATA_SEGMENT_RELRO_END. With it being cleared if there is no DATA_SEGMENT_RELRO_END. Wouldn't this unconditionally add it back in again even if there was no DATA_SEGMENT_RELRO_END?

Reply:
`ScriptParser::readSections` clears `relro` if there is no `DATA_SEGMENT_RELRO_END`.

After `getSectionRank`, `relro` has changed its meaning: whether the section is RELRO, to be checked in `seenRelro |= sec->relro && sec->type != SHT_NOBITS;`.

In the absence of `DATA_SEGMENT_RELRO_END`, this code is still correct.